src/datasets.jl

```diff
@@ -335,14 +335,21 @@ write_dataset(dset::Dataset, memtype::Datatype, x, xfer::DatasetTransferProperti
     API.h5d_write(dset, memtype, API.H5S_ALL, API.H5S_ALL, xfer, x)
 
 # type-specific behaviors
-
+function _check_invalid(dataset::Dataset, buf::AbstractArray)
+    num_bytes_dset = sizeof(datatype(dataset)) * length(dataset)
+    num_bytes_buf = sizeof(eltype(buf)) * length(buf)
```

Comment:
https://docs.julialang.org/en/v1/base/math/#Base.Checked.checked_mul

```julia
julia> Base.checked_mul(typemax(Int)÷2, typemax(Int)-5)
ERROR: OverflowError: 4611686018427387903 * 9223372036854775802 overflowed for type Int64
```

```suggestion
    num_bytes_dset = Base.checked_mul(sizeof(datatype(dataset)), length(dataset))
    num_bytes_buf = Base.checked_mul(sizeof(eltype(buf)), length(buf))
```